### PR TITLE
add loader inference

### PR DESCRIPTION
**Pull request by kentcdodds**
I'm not getting the loader inference I expected. `data` ends up being `any`. Am I doing something wrong here @colinhacks?

### CODE REVIEW

File: app/routes/notes.tsx [hunk 1]
```diff
@@ -6,18 +6,14 @@ import { requireUserId } from "~/session.server";
 import { useUser } from "~/utils";
 import { getNoteListItems } from "~/models/note.server";
 
-type LoaderData = {
-  noteListItems: Awaited<ReturnType<typeof getNoteListItems>>;
-};
-
 export const loader: LoaderFunction = async ({ request }) => {
```
Comment (penspinner) [inline app/routes/notes.tsx:9]:
```suggestion
export const loader = async ({ request }) => {
```
What happens if you remove `LoaderFunction`?

Comment (mcansh) [reply app/routes/notes.tsx:9]: yup, need to type the args and not the full function as LoaderFunction returns `Promise<Response> | Response | Promise<AppData> | AppData`

Comment (kentcdodds) [reply app/routes/notes.tsx:9]: Though I could be convinced. I just don't remember signing up for this so I need to process it a bit...

Comment (MichaelDeBoey) [reply app/routes/notes.tsx:9]: If I remember correctly, somebody said that we would need to choose between automatic type inference & typing the function (with possibly providing the return type)

Comment (kentcdodds) [reply app/routes/notes.tsx:9]: Ok, after looking at it more, I realize that it's required for TypeScript to infer the return type. My mistake. Fixing now.

Comment (kentcdodds) [reply app/routes/notes.tsx]: Even better: https://github.com/microsoft/TypeScript/issues/38511

**Pull request merged**